CoreTiming: ScheduleEvent cleanup

JosJuice opened the pull request:
Now Core::IsCPUThread() only gets called once when using the AnyThread variant. Also, I think the enum approach makes calling code clearer.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/3987)
<!-- Reviewable:end -->

lioncash commented:
Review status: 0 of 26 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/EXI.h, line 8 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KMO5x1BaZmZ9WREYhkF:-KMO5x1C_cgEk6OYxhak:1637347311) ([raw file](https://github.com/dolphin-emu/dolphin/blob/72a94add2f89d2ec77198f63eedaa58ba3fd05a7/Source/Core/Core/HW/EXI.h#L8)):*
> ```objc
> 
> #include "Common/CommonTypes.h"
> #include "Core/CoreTiming.h"
> ```

You can use a forward declaration for FromThread instead of an include, since it's a strongly typed enum.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE.h, line 13 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KMO69R9HGjERUohTv8y:-KMO69R9HGjERUohTv8z:-499318459) ([raw file](https://github.com/dolphin-emu/dolphin/blob/72a94add2f89d2ec77198f63eedaa58ba3fd05a7/Source/Core/Core/IPC_HLE/WII_IPC_HLE.h#L13)):*
> ```objc
> #include "Common/CommonTypes.h"
> 
> #include "Core/CoreTiming.h"
> ```

Ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 26 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/Core/Core/HW/EXI.h, line 8 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KMO5x1BaZmZ9WREYhkF:-KMOECJlO6vL9GjIi5MW:-1156893577) ([raw file](https://github.com/dolphin-emu/dolphin/blob/72a94add2f89d2ec77198f63eedaa58ba3fd05a7/Source/Core/Core/HW/EXI.h#L8)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can use a forward declaration for FromThread instead of an include, since it's a strongly typed enum.
</details>
Thanks, I didn't know that was possible. I've made that change in this PR and will also do the equivalent changes in my DiscIO enum PR.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE.h, line 13 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KMO69R9HGjERUohTv8y:-KMOEMmIScMi-zgCip6a:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/72a94add2f89d2ec77198f63eedaa58ba3fd05a7/Source/Core/Core/IPC_HLE/WII_IPC_HLE.h#L13)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> Ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987)*
<!-- Sent from Reviewable.io -->

phire commented:
1. I'm not sure about the explicit enum for everything, perhaps we should make it the last argument and make it default to CPU.

2. We should be a actually be enforcing this enum, perhaps in debug/fast debug builds only.


Though it would be nice to statically enforce the flags with something like clangtidy. But it's out of the scope of this PR.

delroth commented:
Why can't we just have ScheduleEvent figure out if it's being called from CPU or non CPU thread? (basically just use ANY all the time)

phire commented:
Idealy, ScheduleEvent should never be called from anywhere that isn't the CPU thread. 

Our current exceptions are:
 * events from DSPLLE on a thread
 * events from the GUI (reset button presses, Controller unplugging then replugging)
 * events from the gpu thread (which will go away when we remove dual core mode)
 * network packets from the Broadband Adapter.

Making it automatically compensate when someone accidentally uses it from the wrong thread is not something I want it to do.

JosJuice commented:
> I'm not sure about the explicit enum for everything, perhaps we should make it the last argument and make it default to CPU.

That could be a good idea. I'll change the PR over to using that tomorrow if nobody objects.

> We should be a actually be enforcing this enum, perhaps in debug/fast debug builds only.

The `_assert_msg_` already gives us an annoying pop-up when it's being used wrong. That's how JMC discovered the issue that PR #3989 fixes. Do you want us to enforce it in more ways?

> Making it automatically compensate when someone accidentally uses it from the wrong thread is not something I want it to do.

Agreed. If someone wants to make something non-deterministic, it should be explicit.

degasus commented:
This PR looks good, but it's a bit too big coretiming change to be mixed with two schedule_immediate -> schedule(0) behavior changes. May you split this up and get this merged before?

---

Reviewed 2 of 26 files at r1, 13 of 24 files at r2.
Review status: 15 of 18 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/HW/DSP.cpp, line 468 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoV3OsiwPLEdX0vXXv:-KOoV3OsiwPLEdX0vXXw:b-xyrd54) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4d2df0a8cee1b11a3dca8f41edaafda69ee262dc/Source/Core/Core/HW/DSP.cpp#L468)):*
> ```cpp
> {
>   // TODO: Maybe rethink this? ScheduleEvent_Threadsafe_Immediate has unpredictable timing.
>   CoreTiming::ScheduleEvent_Threadsafe_Immediate(et_GenerateDSPInterrupt, type);
> ```

Please do not mix non-trivial cleanups with emulation code. I agree, dropping immediate is a good idea, but this should be done within a small PR before. Else the behavior of "DSP on CPU thread" will change.

---

*[Source/Core/Core/HW/EXI.h, line 15 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoWjI3TZUWsFaL_vam:-KOoWjI3TZUWsFaL_van:b-axgx6j) ([raw file](https://github.com/dolphin-emu/dolphin/blob/52786e0ec5ad2fda066917c1dd73b3ebca851496/Source/Core/Core/HW/EXI.h#L15)):*
> ```objc
> namespace CoreTiming
> {
> enum class FromThread;
> ```

This is a bit too much forward declarations IMO, but feel free to ignore me.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp, line 115 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoWvFZDlce7vBT38wX:-KOoWvFZDlce7vBT38wY:b-f3tkbw) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4d2df0a8cee1b11a3dca8f41edaafda69ee262dc/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp#L115)):*
> ```cpp
>   Memory::Write_U32(Memory::Read_U32(command_address), command_address + 8);
>   // Generate a reply to the IPC command
>   WII_IPC_HLE_Interface::EnqueueReply_Immediate(command_address);
> ```

This also touches the emulation behavior.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_sdio_slot0.cpp, line 12 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoX11Q4nldlVRmB6Fg:-KOoX11Q4nldlVRmB6Fh:bb7hquh) ([raw file](https://github.com/dolphin-emu/dolphin/blob/52786e0ec5ad2fda066917c1dd73b3ebca851496/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_sdio_slot0.cpp#L12)):*
> ```cpp
> #include "Core/ConfigManager.h"
> #include "Core/Core.h"
> #include "Core/CoreTiming.h"
> ```

huh?

---

*[Source/Core/Core/IPC_HLE/WII_Socket.cpp, line 12 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoX3qq-Rt9MDvoQpi0:-KOoX3qq-Rt9MDvoQpi1:bb7hquh) ([raw file](https://github.com/dolphin-emu/dolphin/blob/52786e0ec5ad2fda066917c1dd73b3ebca851496/Source/Core/Core/IPC_HLE/WII_Socket.cpp#L12)):*
> ```cpp
> #include "Common/FileUtil.h"
> #include "Core/Core.h"
> #include "Core/CoreTiming.h"
> ```

huh?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-:-KOoXLF-QNbvYfW3jxTl:b-ymhtrh)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 10 of 17 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_sdio_slot0.cpp, line 12 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoX11Q4nldlVRmB6Fg:-KOofewKkN9R3XeQz78e:b-3lzsv8) ([raw file](https://github.com/dolphin-emu/dolphin/blob/52786e0ec5ad2fda066917c1dd73b3ebca851496/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_sdio_slot0.cpp#L12)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> huh?
</details>
Must've been from when FromThread always was explicit. Removed.

---

*[Source/Core/Core/IPC_HLE/WII_Socket.cpp, line 12 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoX3qq-Rt9MDvoQpi0:-KOofoAm-05r4jYJ4Cza:bss4a2b) ([raw file](https://github.com/dolphin-emu/dolphin/blob/52786e0ec5ad2fda066917c1dd73b3ebca851496/Source/Core/Core/IPC_HLE/WII_Socket.cpp#L12)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> huh?
</details>
Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987)*
<!-- Sent from Reviewable.io -->

degasus commented:
Reviewed 10 of 14 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[Source/Core/Core/HW/ProcessorInterface.cpp, line 219 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOouznEWgk852S5zmQT:-KOouznEWgk852S5zmQU:b-rbu0gn) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e0c3de292f6fa2c875367d927be773e6605b9b5b/Source/Core/Core/HW/ProcessorInterface.cpp#L219)):*
> ```cpp
> {
>   CoreTiming::ScheduleEvent(0, toggleResetButton, true, CoreTiming::FromThread::ANY);
>   CoreTiming::ScheduleEvent(0, toggleResetButton, 0, CoreTiming::FromThread::ANY);
> ```

Are you sure to call this event twice?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/Core/HW/ProcessorInterface.cpp, line 219 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOouznEWgk852S5zmQT:-KOovQM4zN7JjXSZZ4Bg:bbx6fj3) ([raw file](https://github.com/dolphin-emu/dolphin/blob/e0c3de292f6fa2c875367d927be773e6605b9b5b/Source/Core/Core/HW/ProcessorInterface.cpp#L219)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Are you sure to call this event twice?
</details>
No, it's a copy-paste fail. This one is supposed to be `iosNotifyResetButton`. Thanks for catching it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 12 of 15 files reviewed at latest revision, 3 unresolved discussions.

---

*[Source/Core/Core/HW/DSP.cpp, line 468 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoV3OsiwPLEdX0vXXv:-KOtD1WuSLjWQM1Yp4xG:b-3yg7io) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4d2df0a8cee1b11a3dca8f41edaafda69ee262dc/Source/Core/Core/HW/DSP.cpp#L468)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> Please do not mix non-trivial cleanups with emulation code. I agree, dropping immediate is a good idea, but this should be done within a small PR before. Else the behavior of "DSP on CPU thread" will change.
</details>
Done in a separate PR.

---

*[Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp, line 115 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-KOoWvFZDlce7vBT38wX:-KOtD5CKTFWkc5XFF0w7:b-3yg7io) ([raw file](https://github.com/dolphin-emu/dolphin/blob/4d2df0a8cee1b11a3dca8f41edaafda69ee262dc/Source/Core/Core/IPC_HLE/WII_IPC_HLE_Device_DI.cpp#L115)):*
<details><summary><i>Previously, degasus (Markus Wick) wrote…</i></summary>
> This also touches the emulation behavior.
</details>
Done in a separate PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987)*
<!-- Sent from Reviewable.io -->

degasus commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 4 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3987#-:-KOtELTn0BCrEkFY-OU_:bnfp4nl)*
<!-- Sent from Reviewable.io -->
